vpr/src/route/rr_node.h

```diff
@@ -157,43 +122,21 @@ class t_rr_node {
     void set_direction(e_direction);
     void set_side(e_side);
 
-  private: //Types
-    //The edge information is stored in a structure to economize on the number of pointers held
-    //by t_rr_node (to save memory), and is not exposed externally
-    struct t_rr_edge {
-        int sink_node = -1;   //The ID of the sink RR node associated with this edge
-        short switch_id = -1; //The ID of the switch type this edge represents
-    };
+    void next_node() {
+        id_ = RRNodeId((size_t)(id_) + 1);
+    }
+
+    RRNodeId id() const {
+        return id_;
+    }
+
+    void prev_node() {
+        id_ = RRNodeId((size_t)(id_)-1);
+    }
 
   private: //Data
-    //Note: we use a plain array and use small types for sizes to save space vs std::vector
-    //      (using std::vector's nearly doubles the size of the class)
-    std::unique_ptr<t_rr_edge[]> edges_ = nullptr;
-    t_edge_size num_edges_ = 0;
-    t_edge_size edges_capacity_ = 0;
-    uint8_t num_non_configurable_edges_ = 0;
-
-    int8_t cost_index_ = -1;
-    int16_t rc_index_ = -1;
-
-    int16_t xlow_ = -1;
-    int16_t ylow_ = -1;
-    int16_t xhigh_ = -1;
-    int16_t yhigh_ = -1;
-
-    t_rr_type type_ = NUM_RR_TYPES;
-    union {
-        e_direction direction; //Valid only for CHANX/CHANY
-        e_side side;           //Valid only for IPINs/OPINs
-    } dir_side_;
-
-    union {
-        int16_t pin_num;
-        int16_t track_num;
-        int16_t class_num;
-    } ptc_;
-    t_edge_size fan_in_ = 0;
-    uint16_t capacity_ = 0;
+    t_rr_node_storage* storage_;
+    RRNodeId id_;
```

Comment:
A thought on a potential optimization.

VPR only has a single RRGraph at any point in time. Instead of paying the overhead for:
```
t_rr_node_storage* storage_;
```
in each t_rr_node proxy object, you could just replace it with a call to:
```
g_vpr_ctx.device().rr_nodes;
```
where needed, which would make `sizeof(t_rr_node) == sizeof(RRNodeId)`.

While in the long-term I think we should really move to better encapsulating the RR graph as a whole this would be a temporary optimization.